File: gcpc2025/fairandsquare/submissions/accepted/yidi.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define all(x) ::begin(x), ::end(x)
void _d(auto... x) { ((cerr << ' ' << x), ...) << endl; }
#define debug(x...) cerr << "["#x"]:", _d(x)

int main() {
    cin.tie(0)->sync_with_stdio(0);

    int n, m;
    cin >> n >> m;
    int cnt = 0;
    vector<string> a(n);
    for (string& s : a) cin >> s, cnt += count(all(s), '#');

    int ans = 1;
    for (int x=2; x<=n; x++) {
        if (cnt % (x*x) != 0) continue;

        int cnt2 = cnt / (x*x);
        vector<vector<char>> vis(n, vector<char>(m));
        for (int i=0; i+x<=n; i++) {
            for (int j=0; j+x<=m; j++) if (!vis[i][j] && a[i][j] == '#') {
                cnt2--;
                for (int k=0; k<x; k++) {
                    for (int l=0; l<x; l++) {
                        if (a[i+k][j+l] != '#' || vis[i+k][j+l]) cnt2 = -1;
                        vis[i+k][j+l] = true;
                    }
                }
            }
        }
        if (!cnt2) ans = x;
    }
    cout << ans << '\n';
}

File: gcpc2025/fairandsquare/submissions/accepted/yidi.py
n, m = map(int, input().split())
a = [input() for i in range(n)]

cnt = sum(s.count('#') for s in a)

ans = 1
for x in range(1, n+1):
    if cnt % (x*x) != 0:
        continue

    cnt2 = cnt / x / x
    vis = [[False]*m for i in range(n)]
    for i in range(n-x+1):
        for j in range(m-x+1):
            if not vis[i][j] and a[i][j] == '#':
                cnt2 -= 1
                for k in range(x):
                    for l in range(x):
                        if a[i+k][j+l] != '#' or vis[i+k][j+l]:
                            cnt2 = -1
                        vis[i+k][j+l] = True

    if not cnt2:
        ans = x

print(ans)
